shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/builder/SchemaBuilderWithDefaultLoader.java

```diff
@@ -0,0 +1,132 @@
+import org.apache.shardingsphere.infra.metadata.schema.builder.loader.ColumnMetaDataLoader;
+import org.apache.shardingsphere.infra.metadata.schema.builder.loader.SchemaMetaDataLoader;
+import org.apache.shardingsphere.infra.metadata.schema.builder.loader.adapter.MetaDataLoaderConnectionAdapter;
+import org.apache.shardingsphere.infra.metadata.schema.model.ColumnMetaData;
+import org.apache.shardingsphere.infra.metadata.schema.model.TableMetaData;
+
+import javax.sql.DataSource;
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.stream.Collectors;
+
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public final class SchemaBuilderWithDefaultLoader {
+    
+    /**
+     * Build table metadata with default loader.
+     * @param executorService executorService
+     * @param materials schema builder materials
+     * @param logicTableDataNodesMap map of logicTable to DataNodes
+     * @return metadata map
+     * @throws SQLException SQL exception
+     */
+    public static Map<String, TableMetaData> build(final ExecutorService executorService, final SchemaBuilderMaterials materials,
+            final Map<String, Collection<DataNode>> logicTableDataNodesMap) throws SQLException {
+        Map<String, TableMetaData> result = new HashMap<>(materials.getRules().size(), 1);
+        result.putAll(appendActualTables(executorService, materials, logicTableDataNodesMap));
+        result.putAll(appendRuleLogicTables(executorService, materials, logicTableDataNodesMap));
+        return result;
+    }
+    
+    private static Map<String, TableMetaData> appendActualTables(final ExecutorService executorService, final SchemaBuilderMaterials materials,
+            final Map<String, Collection<DataNode>> logicTableDataNodesMap) throws SQLException {
+        Collection<String> existedTableNames = logicTableDataNodesMap.values().stream()
+                .flatMap(each -> each.stream().map(DataNode::getTableName))
+                .collect(Collectors.toSet());
+
+        Map<String, TableMetaData> result = new HashMap<>(materials.getRules().size());
+        for (Map.Entry<String, DataSource> entry : materials.getDataSourceMap().entrySet()) {
+            Collection<String> tableNames = SchemaMetaDataLoader.loadAllTableNames(entry.getValue(), materials.getDatabaseType());
+            tableNames.removeAll(existedTableNames);
+
+            Collection<Future<TableMetaData>> futures = new ArrayList<>(tableNames.size());
+            for (String each : tableNames) {
+                futures.add(executorService.submit(() -> loadTableMetaData(each, entry.getValue(), materials.getDatabaseType())));
```

Comment:
Each table is individually submitted to the thread pool, can it really improve performance?